actfw_core/unicam_isp_capture.py

```diff
@@ -143,32 +196,90 @@ def produce_image_from_isp(self) -> None:
+
+        return y_sum / pixel_sum / (1 << PIPELINE_BITS)
+
+    def agc(self, stats: bcm2835_isp_stats) -> None:
+        current_y = self.calculate_y(stats) * self.degital_gain  # apply degital gain
+        additional_gain = min(10, self.target_Y / (current_y + 0.001))
+        max_exposure = SHUTTERS[-1] * GAINS[-1]
+        target_exposure = min(self.exposure * additional_gain, max_exposure)
+        analogue_gain = GAINS[0]
+        shutter_time = SHUTTERS[0]
+        if shutter_time * analogue_gain < target_exposure:
+            for stage in range(1, len(GAINS)):
+                max_analogue_gain = GAINS[stage]
+                max_shutter_time = SHUTTERS[stage]
+                # fix gain, increase shutter time
+                if max_shutter_time * analogue_gain >= target_exposure:
+                    shutter_time = target_exposure / analogue_gain
+                    break
+                shutter_time = max_shutter_time
+
+                # fix shutter time, increase gain
+                if shutter_time * max_analogue_gain >= target_exposure:
+                    analogue_gain = target_exposure / shutter_time
+                    break
+                analogue_gain = max_analogue_gain
+
+        # may need flicker avoidance here. ref. https://github.com/kbingham/libcamera/blob/d7415bc4e46fe8aa25a495c79516d9882a35a5aa/src/ipa/raspberrypi/controller/rpi/agc.cpp#L724 # noqa: E501, B950
+
+        # print(f"analogue_gain: {analogue_gain}, shutter: {shutter_time}")
+
+        self.exposure = shutter_time * analogue_gain
+        self.set_unicam_exposure(analogue_gain, shutter_time)
+
+    def awb(self, stats: bcm2835_isp_stats) -> None:
+        sum_r = 0
+        sum_b = 0
+        sum_g = 0
+        # len(stats.awb_stats) == 192
+        for region in stats.awb_stats:
+            sum_r += region.r_sum
+            sum_b += region.b_sum
+            sum_g += region.g_sum
+        self.gain_r = sum_g / (sum_r + 1)
+        self.gain_b = sum_g / (sum_b + 1)
+        red_balance_ctrl = v4l2_ext_control()
+        red_balance_ctrl.id = V4L2_CID.RED_BALANCE
+        red_balance_ctrl.value64 = int(self.gain_r * 1000)
+        blue_balance_ctrl = v4l2_ext_control()
+        blue_balance_ctrl.id = V4L2_CID.BLUE_BALANCE
+        blue_balance_ctrl.value64 = int(self.gain_b * 1000)
+        self.isp_in.set_ext_controls([red_balance_ctrl, blue_balance_ctrl])
+
     def adjust_setting_from_isp(self) -> None:
         buffer = self.isp_out_metadata.dequeue_buffer_nonblocking(v4l2_memory=V4L2_MEMORY.MMAP)
         if buffer is None:
             return
 
         stats: bcm2835_isp_stats = cast(buffer.mapped_buf, POINTER(bcm2835_isp_stats)).contents
+        if self.do_agc:
+            if self.agc_interval_count < AGC_INTERVAL:
+                self.agc_interval_count += 1
+            else:
+                self.agc_interval_count = 0
+                self.agc(stats)
```

Comment:
run agc once in 3 frames because there is a delay between control values being requested and the sensor applying them. 
More ideally, embedded data should be used, but this does not seem possible with imx219.
(ref. https://github.com/kbingham/libcamera/blob/d7415bc4e46fe8aa25a495c79516d9882a35a5aa/src/ipa/raspberrypi/cam_helper_imx219.cpp#L13-L16)